### DISCUSSION TITLE
rewrites safe to use in production

### DISCUSSION THREAD

**Discussion by hongkiulam** [vercel/next.js#59716] [Help] [answered]
Hi 👋,

For context, I am running an old legacy application and looking to slowly recreate the FE in nextjs.

For as long as I am running both apps side by side I need to ensure routing is seamless between the two. I've looked into doing this with the load balancer and nginx, but was happy to come across rewrites which does what I need but with less config.

```
If page exists in Next.js, render it
Otherwise all other requests are handled by legacy app via rewrites
```

I have asked around but no one I know has much experience with rewrites, and it's suggested to be some sort of an anti-pattern, leaning towards keeping this logic in the load balancer.

Does anyone know if it can be used in production?

Does it compare well to the other approaches I mentioned?

Thanks!

**Comment by matthewraaff** [comment, marked as answer]
Hello,

Rewrites this way can be used in production, I've personally been using them in my prod environment in this context for a while and I've not ran in to any issues yet. There are some good [documentation pages](https://nextjs.org/docs/pages/api-reference/next-config-js/rewrites) which can help you with your rewrites like this if you need them.

It is somewhat of an anti-pattern, and if you could use nginx to route it, that'd be seen as a better practise to some people. You could use nginx fallbacks to return the legacy application when the main one fails. [This SO answer](https://stackoverflow.com/a/49870894) has the following nginx config which could be of help:

```nginx
server {
    location / {
        proxy_pass http://new_server;
        error_page 404 500 502 503 504 = @fallback;
    }

    location @fallback {
        proxy_pass http://old_server;
    }
}
```

tldr; other people see it as an anti-pattern, but you're using it in a positive way. nginx routing would be seen as better, but it all depends on the context and how you want to manage the application.

**Comment by hongkiulam** [comment]
Thanks for your response @matthewraaff!

It's great to hear someone else is using this feature this way, and in production. I haven't seen nginx setup like that before so also very interesting, I'll explore both 👍 

For my use case, keeping the response time low while under high load is important - was this also part of your criteria?

**Comment by matthewraaff** [comment]
Good afternoon, @hongkiulam 

My criteria did involve performance, but I found from some (admittedly low level 😂) benchmarking a while back that the nginx and rewrites had very similar performance. If performance is a big factor to you, there are a few extra steps you can take.

OpenLightSpeed is a faster web server, and you could experiment with caching which pages need to fallback to see what works properly. This could also be done with nginx, but I'm not sure how complex it would be, and I haven't done it before.

You could also do this caching within the application; An imaginary setup that'd probably work quite well for having many requests is having your nextjs server reporting which URLs need to fallback as they return an error, and having a cache database, such as Redis, telling your reverse proxies which urls need to return the legacy application and which don't. However, I don't think this would be the best approach unless you have the time and resources to build this.

It's also very dependant on how performance-critical this application is, and in what metric. Are there lots of visitors, do you need a quick load time, time to first byte, connection time, etc.

Personally, I'd stick to something like the config from that SO answer, or try a very similar config on a faster web server and see what happens. 

I hope this response helps!

**Comment by hongkiulam** [comment]
Thanks @matthewraaff,

I'm going to go with `rewrites` in the end, I have it in a staging environment now and will perform some load tests soon
